pil-analyzer/tests/types.rs

```diff
@@ -614,3 +669,88 @@ fn new_fixed_column_wrong_type() {
     "#;
     type_check(input, &[]);
 }
+
+#[test]
+fn impl_type_resolution() {
+    let input = "
+    namespace std::convert(4);
+        let fe = || fe();
+    namespace F(4);
+    
+        trait Add<T> {
+            add: T, T -> T,
+        }
+
+        impl Add<fe> {
+            add: |a, b| a + b,
+        }
+
+        impl Add<int> {
+            add: |a, b| a + b,
+        }
+
+        let r: int = Add::add(3, 4);
+    ";
+    type_check(input, &[("F::r", "", "int")]);
+}
+
+#[test]
+fn trait_multi_generics() {
+    let input = "
+    trait ToTuple<S, I> {
+        get: S -> (S, I),
+    }
+    impl ToTuple<int, (int, int)> {
+        get: |n| (n, (1, n+2)),
+    }
+    let r: (int, (int, int)) = ToTuple::get(3);
+    ";
+    type_check(input, &[("r", "", "(int, (int, int))")]);
+}
+
+#[test]
+fn trait_with_user_defined_enum() {
+    let input = "
+    enum Bool { True, False }
+    
+    trait Not<T> {
+        not: T -> T,
+    }
+    
+    impl Not<Bool> {
+        not: |b| match b {
+            Bool::True => Bool::False,
+            Bool::False => Bool::True,
+        },
+    }
+    let b = Not::not(Bool::True);
+    ";
+    type_check(input, &[("b", "", "Bool")]);
+}
+
+#[test]
+fn trait_with_user_defined_enum2() {
+    let input = "
+    enum V1 { A }
+    enum V2 { B }
+
+    trait Convert<T, U> {
+        convert: T -> U,
+    }
+
+    impl Convert<V1, V2> {
+        convert: |x| match x {
+            V1::A => V2::B,
+        },
+    }
+    impl Convert<V2, V1> {
+        convert: |x| match x {
+            V2::B => V1::A,
+        },
+    }
+
+    let r1: V2 = Convert::convert(V1::A);
```

Comment:
I think to test anything here, you should remove the types of r1 and r2.

Reply:
But then the type checker will complain because it cannot assign types, and it's correct.

Reply:
So I think the test is good, but what we need is a negative test as well: Call `Convert::convert(false)`.